src/code/sys_flashrom.c

```diff
@@ -3,42 +3,242 @@
+    u32 flashType;
+    u32 flashVendor;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    osFlashReadId(&flashType, &flashVendor);
+    sFlashromVendor = flashVendor;
+    if (flashType == 0x11118001) {
+        if ((flashVendor == FLASH_VERSION_MX_PROTO_A) || (flashVendor == FLASH_VERSION_MX_A) ||
+            (flashVendor == FLASH_VERSION_MX_C) || (flashVendor == FLASH_VERSION_MEI) ||
+            (flashVendor == FLASH_VERSION_MX_B_AND_D)) {
+            return 0;
+        } else {
+            return -1;
+        }
+    }
+    return -1;
+}
+
+s32 SysFlashrom_InitFlash(void) {
+    osCreateMesgQueue(&sFlashromMesgQueue, sFlashromMesg, ARRAY_COUNT(sFlashromMesg));
+    osFlashInit();
+    sFlashromFlashIsGood = true;
+    if (SysFlashrom_CheckFlashType()) {
+        sFlashromFlashIsGood = false;
+        return -1;
+    }
+    return 0;
+}
+
+s32 SysFlashrom_ReadData(void* addr, s32 pageNum, s32 pageCount) {
+    OSIoMesg msg;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    osInvalDCache(addr, pageCount * 128);
+    osFlashReadArray(&msg, 0, pageNum, addr, pageCount, &sFlashromMesgQueue);
+    osRecvMesg(&sFlashromMesgQueue, NULL, OS_MESG_BLOCK);
+    return 0;
+}
+
+s32 SysFlashrom_ErasePage(u32 page) {
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    return osFlashSectorErase(page);
+}
+
+s32 SysFlashrom_WriteData(u8* addr, u32 pageNum, u32 pageCount) {
+    OSIoMesg msg;
+    s32 result;
+    u32 i;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    if (pageNum & 0x7F) {
+        Fault_AddHungupAndCrash("../sys_flashrom.c", 275);
+    }
+    osWritebackDCache(addr, pageCount * 128);
+    for (i = 0; i != pageCount; i++) {
+        osFlashWriteBuffer(&msg, 0, addr + i * 128, &sFlashromMesgQueue);
+        osRecvMesg(&sFlashromMesgQueue, NULL, OS_MESG_BLOCK);
+        result = osFlashWriteArray(i + pageNum);
+        if (result != 0) {
+            return result;
+        }
+    }
+    return 0;
+}
+
+s32 SysFlashrom_AttemptWrite(void* addr, s32 pageNum, s32 pageCount) {
+    s32 result;
+    s32 i;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    osWritebackDCache(addr, pageCount * 128);
+    i = 0;
+failRetry:
+    result = SysFlashrom_ErasePage(pageNum);
+    if (result != 0) {
+        if (i < 3) {
+            i++;
+            goto failRetry;
+        }
+        return result;
+    }
+    result = SysFlashrom_WriteData(addr, pageNum, pageCount);
+    if (result != 0) {
+        if (i < 3) {
+            i++;
+            goto failRetry;
+        }
+        return result;
+    }
+    return 0;
+}
+
+s32 func_80185BE4(void* data, void* addr, u32 pageCount) {
+    u32 size;
+    u32 i;
+
+    for (i = 0; i < pageCount * 128; i += 4) {
+        if ((*(s32*)data & *(s32*)addr) != *(s32*)addr) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+s32 SysFlashrom_WriteData2(void* addr, u32 pageNum, u32 pageCount) {
```

Comment:
Since I think the current `SysFlashrom_WriteData` can be renamed, this one could become `SysFlashrom_WriteData`